platform/stm32/export/aux/i2c_bus.hpp

```diff
@@ -0,0 +1,681 @@
+    //! \param[in] handler Handler itself.
+    //!
+    void set_handler(const handler_fn &handler);
+
+    //!
+    //! \brief Reset xfer buffers.
+    //! Buffers that were set by \sa set_tx() and \sa set_rx()
+    //! will be no longer used after this call.
+    //!
+    void reset_buffers();
+
+    //!
+    //! \brief Resets previously set handler.
+    //!
+    void reset_handler();
+
+    //!
+    //! \brief Executes xfer, using buffers previously set.
+    //! When it will be done, handler will be invoked.
+    //! \return Status of operation.
+    //!
+    ecl::err do_xfer();
+
+    //!
+    //! \brief Sets slave address.
+    //! This address is used during communication.
+    //! In case there is only one device on the bus,
+    //! this function may be called only once,
+    //! before first xfer().
+    void set_slave_addr(uint16_t addr);
+
+private:
+    static constexpr auto pick_rcc();
+    static constexpr auto pick_rcc_fn();
+    static constexpr auto pick_i2c();
+    static constexpr auto pick_ev_irqn();
+    static constexpr auto pick_er_irqn();
+
+    //! Communication direction
+    // TODO Add slave support
+    enum direction {
+        MASTER_TX,
+        MASTER_RX,
+    } m_direction;
+
+    // Device status flags
+
+    //! Bit set if device initialized.
+    static constexpr uint8_t m_inited     = 0x1;
+
+    // Device status methods
+    inline bool inited() const     { return (m_status & m_inited)  != 0; }
+
+    inline void set_inited()       { m_status |= m_inited; }
+    inline void clear_inited()     { m_status &= ~(m_inited); }
+
+    //! Setup xfer in IRQ mode
+    ecl::err i2c_setup_xfer_irq();
+
+    //! Operations for POLL mode
+    ecl::err i2c_transmit_poll();
+    ecl::err i2c_receive_poll();
+
+    //! Handles IRQ events from a bus.
+    void irq_ev_handler();
+    //! Handles IRQ events (error) from a bus.
+    void irq_er_handler();
+
+    //! Helper functions for bytes receive/transmit
+    void receive_bytes(size_t count);
+    void send_bytes(size_t count);
+
+    handler_fn      m_event_handler; //! Handler passed via set_handler().
+    const uint8_t   *m_tx;           //! Transmit buffer.
+    size_t          m_tx_size;       //! TX buffer size.
+    size_t          m_tx_left;       //! Left to send in TX buffer.
+    uint8_t         *m_rx;           //! Receive buffer.
+    size_t          m_rx_size;       //! RX buffer size.
+    size_t          m_rx_left;       //! Left to receive in RX buffer.
+    uint8_t         m_status;        //! Tracks device status.
+
+    uint16_t        m_own_addr;     //! device address in slave mode
+    uint16_t        m_slave_addr;   //! address of slave to communicate
+};
+
+template< class i2c_config >
+i2c_bus< i2c_config >::i2c_bus()
+	:m_event_handler{}
+	,m_tx{nullptr}
+	,m_tx_size{0}
+	,m_tx_left{0}
+	,m_rx{nullptr}
+	,m_rx_size{0}
+	,m_rx_left{0}
+	,m_status{0}
+	,m_slave_addr{0}
+{
+
+}
+
+template< class i2c_config >
+i2c_bus< i2c_config >::~i2c_bus()
```

Comment:
Basic types have no default constructors, so those sub-object will not be initialized properly by default.